x-pack/plugins/enterprise_search/public/applications/enterprise_search_content/components/search_index/pipelines/ml_inference/text_expansion_callout.tsx

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import React, { useCallback, useEffect, useMemo, useState } from 'react';
+
+import { useValues } from 'kea';
+
+import {
+  EuiBadge,
+  EuiButtonIcon,
+  EuiFlexGroup,
+  EuiFlexItem,
+  EuiLink,
+  EuiPanel,
+  EuiText,
+  EuiTitle,
+} from '@elastic/eui';
+import { i18n } from '@kbn/i18n';
+import { FormattedMessage, FormattedHTMLMessage } from '@kbn/i18n-react';
+
+import { docLinks } from '../../../../../shared/doc_links';
+
+import { MLInferenceLogic } from './ml_inference_logic';
+
+export interface TextExpansionCallOutState {
+  dismiss: () => void;
+  dismissable: boolean;
+  show: boolean;
+}
+
+export interface TextExpansionCallOutProps {
+  dismissable?: boolean;
+}
+
+export const TEXT_EXPANSION_CALL_OUT_DISMISSED_KEY =
+  'enterprise-search-text-expansion-callout-dismissed';
+
+export const useTextExpansionCallOutData = ({
+  dismissable = false,
+}: TextExpansionCallOutProps): TextExpansionCallOutState => {
+  const { supportedMLModels } = useValues(MLInferenceLogic);
+
+  const doesNotHaveTextExpansionModel = useMemo(() => {
+    return !supportedMLModels.some((m) => m.inference_config?.text_expansion);
+  }, [supportedMLModels]);
+
+  const [show, setShow] = useState<boolean>(() => {
+    if (!dismissable) return true;
+
+    try {
+      return localStorage.getItem(TEXT_EXPANSION_CALL_OUT_DISMISSED_KEY) !== 'true';
+    } catch {
+      return true;
+    }
+  });
+
+  useEffect(() => {
+    try {
+      localStorage.setItem(TEXT_EXPANSION_CALL_OUT_DISMISSED_KEY, JSON.stringify(!show));
+    } catch {
+      return;
+    }
+  }, [show]);
+
+  const dismiss = useCallback(() => {
```

Comment:
nitpick (and a question): is wrapping this function in useCallback doing anything? useCallback makes things a little less readable and in most cases doesn't actually add any performance benefits (see: https://react.dev/reference/react/useCallback).

Reply:
it does technically only allocate a single function for each `TextExpansionCallOut`. because i pulled all of this logic into a separate hook it's just habit i guess.

> If you’re writing a custom Hook, it’s recommended to wrap any functions that it returns into `useCallback`
> https://react.dev/reference/react/useCallback#optimizing-a-custom-hook

because this hook is only ever used in this file i guess that rule doesn't really apply. but i also really doubt anyone would have a hard time reading this code.